### DISCUSSION TITLE
How to use magit rebase exec?

### DISCUSSION THREAD

**Discussion by mattiasdrp** [magit/magit#5127] [Support] [answered]
If I type `git rebase -x 'echo "toto"' HEAD~5` in my terminal I will have the following output:

```
❯ git rebase -x 'echo "toto"' HEAD~5
Executing: echo "toto"
toto
Executing: echo "toto"
toto
Executing: echo "toto"
toto
Executing: echo "toto"
toto
Executing: echo "toto"
toto
Successfully rebased and updated refs/heads/main.

```

In magit, I put the cursor on HEAD~5, hit `r i` and I have the following buffer:

```
pick ad3667f (...)
pick a0aece6 (...)
pick 2dbef2d (...)
pick bddcb1e (...)
pick 2243232 (...)
```

If I hit `x` to give a shell command, it will look like this:

```
pick ad3667f (...)
exec echo "toto"
pick a0aece6 (...)
pick 2dbef2d (...)
pick bddcb1e (...)
pick 2243232 (...)
```

And the output will be:

```
  0 git … rebase -i --autostash ad3667f\^
hint: Waiting for your editor to close the file...
Executing: echo "toto"
toto
Successfully rebased and updated refs/heads/main.
```

However, I wanted to have the exec command apply to all commits.

I tried selecting them all, `x` into `echo "toto"`, it puts the exec command at the end and no output.

I tried adding it manually to each commit:

```
pick ad3667f (...)
exec echo "toto"
pick a0aece6 (...)
exec echo "toto"
pick 2dbef2d (...)
exec echo "toto"
pick bddcb1e (...)
exec echo "toto"
pick 2243232 (...)
exec echo "toto"
```

But the output is the following:

```
  0 git … rebase -i --autostash ad3667f\^
hint: Waiting for your editor to close the file...
Executing: echo "toto"
toto
Successfully rebased and updated refs/heads/main.
```

Maybe I misunderstood how to use the `exec` feature in magit but I can't make it behave like `--exec` when using git in my terminal.

Thanks for your patience and understanding :-)

**Comment by glasserc** [comment]
I don't think `magit` supports the `-x` argument to `rebase`. As you've discovered, the `x` command in the rebase buffer is more like the `exec` command in a typical interactive rebase. The flag `-x` would probably be something you would specify after pressing `r` (like other flags `-k` and `-r`). You might try implementing it yourself if it's something you use regularly! (I just did that for my pet feature https://github.com/magit/magit/discussions/4822.) Even if you can't get it upstreamed, it might still be a simple enough change that you could maintain it yourself.

Out of curiosity, what do you use `-x` for?

**Comment by mattiasdrp** [comment]
Thanks for your answer.

I use it to compile between every commits when I edit them ;-)

I'll look into implementing it for myself, thanks :-)

**Comment by kyleam** [comment, marked as answer]
The rebase transient has had the `--exec` option since 9cbaf48c (magit-rebase: Add --exec argument, 2020-02-26).  It's hidden by default though.  See `(info "(transient)Enabling and Disabling Suffixes")` ([web][]) for how to enable.

To mimic the initial example you gave (`git rebase -x 'echo "toto"' HEAD~5`), you wouldn't want an interactive rebase. Instead you' could use the suffix `e` and enter "HEAD~5".

Having played around with this a bit, it does look like some of the output isn't getting into the process buffer.  I haven't dug deeper to try to figure out what's going on there, but you can convince yourself that each command is running by, say, appending to file (`--exec=echo . >>foo`).

[web]: https://magit.vc/manual/transient/Enabling-and-Disabling-Suffixes.html
